## API : 
        Your code -->Api call-->Server-->Json-->Your COde
    1. URL : restaurant Menu .
    2. GET : Menu
    3. Post : Place an order
    4. Status Code : Order sucess/order failed

### Basic Structure of api :
                            1. Endpoint url : The adress of the service .
                            2. Http Method : Get,Post(Put ,Patch,Delete)
                            3. Headers->Metadata such as authentication .
                            4. Paarameter or body:
                                                    a. query parameter
                                                    b. Path Parameter
                                                    c. Body

In [20]:
import requests

response=requests.get("https://api.github.com/")

print(response.text)

print(response.status_code)

{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_searc

response.text : gives the simple text form of the api response
reponse.status_code : Tells about weather the fetching being successfull or not .
requests.get: get the data from url .
(200 is the success message .)

Tring to give the wring url 


In [21]:
response=requests.get("https://api.github.com/invalid")
print(response.text)
print(response.status_code)

{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest",
  "status": "404"
}
404


stattus_code : 404 ,
and printed the not found response

In [22]:
response=requests.get("https://api.github.com/")

data=response.json()
print(data["current_user_url"])


https://api.github.com/user


In [23]:
parameters={
    "q":"Agentic",
    "sort":"stars",
    "per_page":4,
    "order":"desc"

}

response=requests.get("https://api.github.com/search/repositories",params=parameters)
data=response.json()
print(data["items"])
print("-"*50)
for i in data["items"]:
    print(i["full_name"])

[{'id': 599320067, 'node_id': 'R_kgDOI7jmAw', 'name': 'langflow', 'full_name': 'langflow-ai/langflow', 'private': False, 'owner': {'login': 'langflow-ai', 'id': 85702467, 'node_id': 'MDEyOk9yZ2FuaXphdGlvbjg1NzAyNDY3', 'avatar_url': 'https://avatars.githubusercontent.com/u/85702467?v=4', 'gravatar_id': '', 'url': 'https://api.github.com/users/langflow-ai', 'html_url': 'https://github.com/langflow-ai', 'followers_url': 'https://api.github.com/users/langflow-ai/followers', 'following_url': 'https://api.github.com/users/langflow-ai/following{/other_user}', 'gists_url': 'https://api.github.com/users/langflow-ai/gists{/gist_id}', 'starred_url': 'https://api.github.com/users/langflow-ai/starred{/owner}{/repo}', 'subscriptions_url': 'https://api.github.com/users/langflow-ai/subscriptions', 'organizations_url': 'https://api.github.com/users/langflow-ai/orgs', 'repos_url': 'https://api.github.com/users/langflow-ai/repos', 'events_url': 'https://api.github.com/users/langflow-ai/events{/privacy}',

### Post : 

In [28]:
response=requests.get("https://jsonplaceholder.typicode.com/posts")
body={
    "name":"Yuvraj Singh",
    "Role":"AI Engineer"
}

response=requests.post("https://jsonplaceholder.typicode.com/posts",json=body)
print(response.json())

{'name': 'Yuvraj Singh', 'Role': 'AI Engineer', 'id': 101}


Now as we post some of our data to the api , then how  can we get it using the id that was provided automatically by api . Let us find out .

In [31]:
post_id=1

reponse=requests.get(f"https://jsonplaceholder.typicode.com/posts/{post_id}")
print(reponse.json())
print(response.status_code)

{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
201


In [33]:


repons=requests.get(f"https://jsonplaceholder.typicode.com/posts/101")
print(repons.json())
print(repons.status_code)

{}
404
